rofl-appd/src/services/metadata.rs

```diff
@@ -0,0 +1,106 @@
+use std::collections::BTreeMap;
+
+use rocket::async_trait;
+use tokio::sync::RwLock;
+
+use oasis_runtime_sdk::modules::rofl::types::{AppInstanceQuery, Registration};
+use rofl_app_core::prelude::*;
+
+/// Namespace prefix for user-provided metadata keys in on-chain registration.
+pub const METADATA_NAMESPACE: &str = "net.oasis.app";
+
+/// An instance metadata service.
+///
+/// User-provided metadata is stored without prefixes and automatically
+/// namespaced with "net.oasis.app." when included in on-chain registration.
+#[async_trait]
+pub trait MetadataService: Send + Sync {
+    /// Set metadata.
+    ///
+    /// This replaces all existing app provided metadata.
+    async fn set(&self, metadata: BTreeMap<String, String>) -> Result<(), Error>;
+
+    /// Get all user-set metadata key-value pairs.
+    async fn get(&self) -> Result<BTreeMap<String, String>, Error>;
+}
+
+/// Error returned by the metadata service.
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("invalid argument")]
+    InvalidArgument,
+
+    #[error(transparent)]
+    Other(#[from] anyhow::Error),
+}
+
+/// A service backed by the Oasis runtime.
+pub struct OasisMetadataService<A: App> {
+    env: Environment<A>,
+    metadata: RwLock<BTreeMap<String, String>>,
+}
+
+impl<A: App> OasisMetadataService<A> {
+    /// Create a new metadata service, initializing from existing on-chain registration.
+    pub async fn new(env: Environment<A>) -> Result<Self, Error> {
+        // Fetch existing metadata from on-chain registration. This ensures that if the app
+        // restarts, previously set metadata is preserved rather than being reset to empty.
```

Comment:
But if the entire app restarts it will have a different RAK, thus the metadata will be empty anyway (since metadata is per app instance, not per app)?